### ISSUE TITLE
How to set persistent event data?

### ISSUE THREAD

**Issue by danepowell** [zumba/amplitude-php#17]
It's awesome that we can call methods like setUserId and setDeviceId to set those properties in a way that persists across events. However, it would be nice if we could persistently set _any_ event property such as os_name or app_version. If I understand correctly, it seems like that's not supported right now, and a downstream project would need to write an event factory or something to ensure that these get set on every event?

**Comment by jonyo** [comment]
>  It's awesome that we can call methods like setUserId and setDeviceId to set those properties in a way that persists across events.

To be clear, it is set per page load / instance of the library (just in case you did not realize what persist refers to, it will not persist between 2 page loads).

> However, it would be nice if we could persistently set any event property such as os_name or app_version.

`os_name` is associated with the user data, so you only need to send once and after that it will be linked to the device / user id until you send an event that changes it.  I thought the same for `app_version` but it is not marked as such, but it does refer to it as user data in other spots in the amplitude docs, you might experiment to see if that info ends up persisting in the use / device data on amplitude's side.

Typically if there is some data you want to persist you would set it as user data.  Note some of the built in fields are automatically treated as user data, see the ones marked w/ footnote of `1` in the docs [here](https://amplitude.zendesk.com/hc/en-us/articles/204771828#keys-for-the-event-argument)

I'm not sure if adding that ability into the library would be wise, as it would add to confusion around the use of user data which only needs to be sent once or when it changes.

**TL;DR:** From your description it sounds like you are persisting data linked to the user or device, in which case there is no need to persist it on your end, amplitude already persists it on their end.

**Comment by danepowell** [comment]
Okay thanks for the detailed response, I'll see if this is something that Amplitude saves on their end!
